tests/store/artifact/test_databricks_artifact_repo.py

```diff
@@ -274,6 +275,89 @@ def test_log_artifact_azure_blob_client_sas_error(self, databricks_artifact_repo
                 databricks_artifact_repo.log_artifact(test_file.strpath)
             write_credential_infos_mock.assert_called_with(run_id=MOCK_RUN_ID, paths=ANY)
 
+    @pytest.mark.parametrize(
+        ("artifact_path", "expected_location"),
+        [(None, "test.txt"), ("output", "output/test.txt"), ("", "test.txt")],
+    )
+    def test_log_artifact_adl_gen2(
+        self, databricks_artifact_repo, test_file, artifact_path, expected_location
+    ):
+        with mock.patch(
+            DATABRICKS_ARTIFACT_REPOSITORY + "._get_write_credential_infos"
+        ) as write_credential_infos_mock, mock.patch(
+            DATABRICKS_ARTIFACT_REPOSITORY + "._azure_adls_gen2_upload_file"
+        ) as azure_upload_mock:
+            mock_credential_info = ArtifactCredentialInfo(
+                signed_uri=MOCK_ADL_GEN2_SIGNED_URI,
+                type=ArtifactCredentialType.AZURE_ADLS_GEN2_SAS_URI,
+            )
+            write_credential_infos_mock.return_value = [mock_credential_info]
+            azure_upload_mock.return_value = None
+            databricks_artifact_repo.log_artifact(test_file.strpath, artifact_path)
+            write_credential_infos_mock.assert_called_with(
+                run_id=MOCK_RUN_ID, paths=[expected_location]
+            )
+            azure_upload_mock.assert_called_with(
+                mock_credential_info, test_file.strpath, expected_location
+            )
+
+    @pytest.mark.parametrize(("artifact_path", "expected_location"), [(None, "test.txt")])
+    def test_log_artifact_adl_gen2_with_headers(
+        self, databricks_artifact_repo, test_file, artifact_path, expected_location
+    ):
+        mock_azure_headers = {
+            "x-ms-content-type": "test-type",
+            "x-ms-owner": "some-owner",
+            "x-ms-blob-type": "some-type",
+        }
+        filtered_azure_headers = {
+            "x-ms-content-type": "test-type",
+            "x-ms-owner": "some-owner",
+        }
```

Comment:
This seems to be testing the existing header removal logic for Gen 1 URLs. Shall we add tests for Gen 2 headers handling? cc @kriscon-db

Reply:
It seems that we don't have special handling logic for Gen 2 headers. The current test should be sufficient.

Reply:
This is for the v2 header. It's different from v1.

Reply:
We don't use header filtering logic for Gen 2 URLs.  The `x-ms-blob-type` header is only added for Gen 1 URLs. See https://src.dev.databricks.com/databricks/universe@229ded1437202b91079a80fb7867b5e8e7fee983/-/blob/common/cloudstorage/MultiCloudPresignedUrlGenerator.scala?L108. This header is explicitly added by `MultiCloudPresignedUrlGenerator`.

I think you can just use:
```python
        mock_azure_headers = {
            "x-ms-encryption-key": "test-key",
            "x-ms-encryption-algorithm": "AES256",
        }
```
and remove `filtered_azure_headers`.

Reply:
I'm explicitly testing these lines of code 
https://github.com/mlflow/mlflow/pull/6730/files#diff-efd79f313ceede37a7b002e5bfcb679e7eecab3451cc5a16eae5c635f9334336R30-R33
It's client side header filtering logic and not in the context of production put/patch requests.

Reply:
I see. So actually the issue is that the header `"x-ms-blob-type"` is specially handled for Gen1 URLs (see details below). To avoid further confusion, I would suggest not mentioning this header in the mock date and using:
```
mock_adl2_headers = { 
    "x-ms-encryption-key": "test-key", 
    "x-ms-encryption-algorithm": "AES256", 
    "non-adl2-header": "some-value"
}

filtered_adl2_headers = {
    "x-ms-encryption-key": "test-key", 
    "x-ms-encryption-algorithm": "AES256",
 }
```
**Details:**

1. CP [adds](https://src.dev.databricks.com/databricks/universe@70d7bb8052a714be0512369ede75a7c1c9ba56e7/-/blob/common/cloudstorage/MultiCloudPresignedUrlGenerator.scala?L108:14) `"x-ms-blob-type"` header explicitly in the presigned URL generator.
2. MLflow Backend immediately [removes this header](https://src.dev.databricks.com/databricks/universe@0a13390b0deb4a0f05fbe29f899dd9f09faf4c3e/-/blob/mlflow/artifacts/main/MlflowArtifactManager.scala?L281) due to MLflow client incompatibility issue.
3. MLflow Client (current) [again removes this header](https://src.dev.databricks.com/mlflow/mlflow@786ecc41704e9a7dbd07e9801c2af14e3e89a782/-/blob/mlflow/azure/client.py?L36) (which shouldn't exist) and [adds this header back](https://src.dev.databricks.com/mlflow/mlflow@786ecc41704e9a7dbd07e9801c2af14e3e89a782/-/blob/mlflow/azure/client.py?L31) before sending the `put_block` request.

The removal of `"x-ms-blob-type"` header already happens in the [backend](https://src.dev.databricks.com/databricks/universe@0a13390b0deb4a0f05fbe29f899dd9f09faf4c3e/-/blob/mlflow/artifacts/main/MlflowArtifactManager.scala?L281). So in the real world, this header shouldn't appear in the return value of `_get_write_credential_infos`.

Reply:
Sure. I was testing sometime supported by V1 but not by V2.